File: samples/system/python/01_basic_chat.py
"""
01 - Basic Chat Completion
==========================

Simplest possible ModelMesh Lite setup: a single provider, one model,
and a chat completion request.

This sample demonstrates:
  - Initializing ModelMesh from an inline configuration dict
  - Obtaining the OpenAI-compatible client
  - Making a chat completion call using a virtual model name
  - Printing the response

The virtual model name "text-generation" maps to a capability pool.  ModelMesh
resolves it to the best (and in this case only) active real model and provider,
then forwards the request transparently.

Uses a mock provider so it runs without API keys.
"""

import asyncio
from modelmesh import ModelMesh, MeshConfig
from modelmesh.interfaces.provider import (
    ChatMessage, CompletionChoice, CompletionRequest, CompletionResponse,
    ErrorClassification, ModelInfo, ModelPricing, ProviderConnector,
    QuotaStatus, RateLimitStatus, TokenUsage,
)


class MockOpenAIProvider(ProviderConnector):
    """Mock provider simulating an OpenAI-style chat model."""

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        return CompletionResponse(
            id="mock-01",
            model="gpt-4o-mini",
            choices=[CompletionChoice(
                index=0,
                message=ChatMessage(
                    role="assistant",
                    content=(
                        "An API gateway is a server that acts as a single entry "
                        "point for a set of microservices, handling request "
                        "routing, authentication, and rate limiting. It simplifies "
                        "client interaction by abstracting the complexity of the "
                        "backend architecture behind a unified interface."
                    ),
                ),
                finish_reason="stop",
            )],
            usage=TokenUsage(prompt_tokens=25, completion_tokens=45, total_tokens=70),
        )

    async def stream(self, request):
        yield await self.complete(request)

    def get_capabilities(self):
        return ["generation.text-generation.chat-completion"]

    def supports(self, cap):
        return cap in self.get_capabilities()

    def list_models(self):
        return [ModelInfo(id="gpt-4o-mini", name="GPT-4o Mini (mock)")]

    def get_model_info(self, mid):
        return ModelInfo(id=mid, name=mid)

    def check_quota(self):
        return QuotaStatus()

    def get_rate_limits(self):
        return RateLimitStatus()

    def get_pricing(self, mid):
        return ModelPricing()

    def report_usage(self, mid, usage):
        pass

    def classify_error(self, error):
        return ErrorClassification(retryable=False)


async def main() -> None:
    # 1. Build a MeshConfig with the mock provider.
    config = MeshConfig(raw={
        "providers": {
            "openai.llm.v1": {
                "connector": "openai.llm.v1",
                "enabled": True,
                "instance": MockOpenAIProvider(),
            },
        },
        "models": {
            "gpt-4o-mini": {
                "provider": "openai.llm.v1",
                "capabilities": [
                    "generation.text-generation.chat-completion",
                ],
            },
        },
        "pools": {
            "text-generation": {
                "capability": "generation.text-generation.chat-completion",
                "strategy": "stick-until-failure",
            },
        },
    })
    mesh = ModelMesh()
    mesh.initialize(config)

    print("ModelMesh initialized with a single provider.\n")

    # 2. Get the OpenAI-compatible client.
    client = mesh.get_client()

    # 3. Make a chat completion request.
    print("Sending chat completion request...")
    response = client.chat.completions.create(
        model="text-generation",
        messages=[
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Explain what an API gateway is in two sentences."},
        ],
        temperature=0.7,
        max_tokens=256,
    )

    # 4. Print the result.
    print(f"\nModel used : {response.model}")
    print(f"Tokens in  : {response.usage.prompt_tokens}")
    print(f"Tokens out : {response.usage.completion_tokens}")
    print(f"\nAssistant:\n{response.choices[0].message.content}")

    # 5. Shut down gracefully.
    mesh.shutdown()
    print("\nModelMesh shut down.")


if __name__ == "__main__":
    asyncio.run(main())

File: samples/system/typescript/01_basicChat.ts
/**
 * 01 - Basic Chat Completion
 *
 * Simplest possible ModelMesh Lite setup: a single OpenAI provider, one model,
 * and a chat completion request. This sample demonstrates:
 *
 *   - Initializing ModelMesh from an inline configuration object
 *   - Obtaining the OpenAI-compatible client
 *   - Making a chat completion call using a virtual model name
 *   - Printing the response
 *
 * The virtual model name "text-generation" maps to the predefined capability
 * pool. ModelMesh resolves it to the best active real model (here, the only
 * configured model from OpenAI) and routes the request transparently.
 *
 * Prerequisites:
 *   - Set the OPENAI_API_KEY environment variable
 */

import { ModelMesh, MeshConfig, CompletionResponse } from "@nistrapa/modelmesh-core";

async function main(): Promise<void> {
  // -----------------------------------------------------------------------
  // 1. Define inline configuration
  // -----------------------------------------------------------------------
  // Secrets are read from environment variables by default
  // (secret-store.modelmesh.env.v1). The provider connector ID follows the
  // naming convention: vendor.service.version (the "provider." prefix is
  // omitted inside the providers section).
  const config = new MeshConfig({
    providers: {
      "openai.llm.v1": {
        enabled: true,
        api_key: "${secrets:OPENAI_API_KEY}",
      },
    },
    // No explicit pool configuration needed -- the predefined
    // "text-generation" pool automatically includes all models registered
    // at the generation.text-generation capability node or its descendants.
    // The default selection strategy is stick-until-failure.
  });

  // -----------------------------------------------------------------------
  // 2. Initialize ModelMesh
  // -----------------------------------------------------------------------
  const mesh = new ModelMesh();
  mesh.initialize(config);
  console.log("ModelMesh initialized with OpenAI provider.");

  // -----------------------------------------------------------------------
  // 3. Get the OpenAI-compatible client
  // -----------------------------------------------------------------------
  // The client exposes the same interface as the official OpenAI SDK.
  // Applications use it as a drop-in replacement.
  const client = mesh.getClient();

  // -----------------------------------------------------------------------
  // 4. Make a chat completion request
  // -----------------------------------------------------------------------
  // The model parameter is a *virtual model name* that maps to a capability
  // pool. "text-generation" resolves to the best active model in the
  // text-generation pool (in this case, a GPT model from OpenAI).
  console.log("Sending chat completion request...");

  const response = (await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Explain what ModelMesh Lite does in one sentence." },
    ],
    temperature: 0.7,
    maxTokens: 150,
  })) as CompletionResponse;

  // -----------------------------------------------------------------------
  // 5. Print the response
  // -----------------------------------------------------------------------
  console.log("\n--- Response ---");
  console.log(`Model used : ${response.model}`);
  console.log(`Response ID: ${response.id}`);
  console.log(`Content    : ${response.choices[0].message?.content}`);
  console.log(`Tokens     : prompt=${response.usage.promptTokens}, completion=${response.usage.completionTokens}`);

  // -----------------------------------------------------------------------
  // 6. Shut down gracefully
  // -----------------------------------------------------------------------
  mesh.shutdown();
  console.log("\nModelMesh shut down.");
}

main().catch(console.error);
